### ISSUE TITLE
Symlinked node_modules directory confuses dts-bundle-generator

### ISSUE THREAD

**Issue by rhuitl** [timocov/dts-bundle-generator#39]
I have to admit, this is probably a quite uncommon use case, but still wanted to report it because maybe it can be fixed easily.

This is based on the same example as #37. Rename your node_modules directory and symlink it:
```
mv node_modules .node_modules_develop
ln -s .node_modules_develop node_modules
```
Then extract the typings. You will get a huge definitions file with a lot of inlined types. It seems like the symlink is confusing the logic that figures out how to deal with a dependency.

I've been using the symlink for a long time now as a technique to make it easy and fast to switch between the modules required for my develop and release branches. It hasn't caused any problems with NPM itself or other tools so far.

**Comment by timocov** [comment]
Hi! It is interesting case. Let me take a look at it, but I'm not sure whether there is easy fix because I get filenames from TypeScript compiler and we do not know whether there were symlinks to file/folder or not.

**Comment by timocov** [comment]
After looking over the issue I have the next conclusions:

1. Indeed, as I said before, I get filenames from TypeScript compiler and I cannot find there the "original path" (path which is requested to resolve).
1. Seems that I found a simple way to fix it - we need to just pass `undefined` or `(path) => path` as `CompilerHost.realpath` to the TypeScript. But I do not know what additional can happen in this case.

I'll add this "fix" under cli option (experimental actually), but let's wait @andy-ms in https://github.com/Microsoft/TypeScript/pull/12020#issuecomment-394087276 first - it is possible (I hope) that he remember what this can lead to.

**Comment by rhuitl** [comment]
Sounds like they added resolving of symlinks to support an `npm link` based workflow. That would break if you set that option and is probably the more important scenario to support out of the box. Maybe you can add the `realpath` argument to the CLI so I can add it to my `dts-bundle-generator` call?

**Comment by timocov** [comment]
Sorry for long delay.

Please check #41 and confirm that it fix your problem. There I have added new CLI option to disable resolving symlinks (`--disable-symlinks-following` - resolving is enabled by default, so I believe it shouldn't break anything else).

I guess that with enabling this option we can get increased memory usage (because multiple files from `node_modules` can be one file, but treated as different files), increased working time (because in some scenarios we need to read/check/etc more files) and possibly some other unexpected errors/behavior. But I believe that it should work in common cases where you have symlinked `node_modules` folder or symlinked packages via `npm link`.

**Comment by timocov** [comment]
https://github.com/timocov/dts-bundle-generator/releases/tag/v1.4.0

**Comment by rhuitl** [comment]
After some testing I think this is not (completely) solved. The typings file has the standard ES typings in it when I used a symlinked node_modules directory (e.g. node_modules -> foobar):

```
Input source files:
  /path/foobar/typescript/lib/lib.dom.d.ts
  /path/foobar/typescript/lib/lib.es2015.d.ts
  /path/foobar/typescript/lib/lib.es5.d.ts
  /path/foobar/typescript/lib/lib.es2015.symbol.wellknown.d.ts
  /path/foobar/typescript/lib/lib.es2015.reflect.d.ts
  /path/foobar/typescript/lib/lib.es2015.proxy.d.ts
  /path/foobar/typescript/lib/lib.es2015.iterable.d.ts
  /path/foobar/typescript/lib/lib.es2015.symbol.d.ts
  /path/foobar/typescript/lib/lib.es2015.promise.d.ts
  /path/foobar/typescript/lib/lib.es2015.generator.d.ts
  /path/foobar/typescript/lib/lib.es2015.collection.d.ts
  /path/foobar/typescript/lib/lib.es2015.core.d.ts
  app.d.ts
```

I thought it was fixed because it magically works when you use ".node_modules" (what I used to test) instead of "foobar" or ".node_modules_develop".

**Comment by timocov** [comment]
Try to add symlinked folder ("foobar" or ".node_modules_develop") to exclude list in tsconfig.

**Comment by timocov** [comment]
Could you please provide new steps to reproduce to be sure that we talk about the same things?

**Comment by rhuitl** [comment]
Yes, of course. Create a new directory. Place these files:

### app.ts
```typescript
export function test(): void { }
```

### package.json
```json
{
  "name": "dts-bundle-generator-bugreport",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "build": "tsc",
    "typings": "dts-bundle-generator --verbose --disable-symlinks-following -o app.d.ts app.ts"
  },
  "dependencies": {},
  "devDependencies": {
    "dts-bundle-generator": "^1.5.0",
    "typescript": "^2.9.1"
  }
}
```

### tsconfig.json
```json
{
  "compilerOptions": {
    "target": "es5",
    "declaration": true,
    "lib": ["es2015", "dom"]
  },
  "files": ["app.ts"]
}
```

Run `npm install`
Run `npm run typings`, look at console output and the generated typings file which look like they should.

Move the `node_modules` directory and symlink it:
```bash
mv node_modules foobar
ln -s foobar node_modules
```

Run `npm run typings` again, this time the console has a lot of output and the typings file contains typings from `es2015` and `dom`.

**Comment by timocov** [comment]
Thank you. I found the problem.

TypeScript compiler does not use `CompilerHost.realpath` to get the path to the lib files. Instead, it uses `CompilerHost.getDefaultLibFileName` and `CompilerHost.getDefaultLibLocation`.

Currently, in `dts-bundle-generator` I use regex to detect that some file is TypeScript's lib file (see [here](https://github.com/timocov/dts-bundle-generator/blob/95ab4d8c954a5fb0bb838b5fd0797142b12fd10b/src/node-modules-helpers.ts#L32-L34)). It is not entirely correct, but it seems that compiler does not provide another way to detect it.

I found in TypeScript's code function `Program.isSourceFileDefaultLibrary` which does exactly I need to detect lib files, but it is not public for now.

I have created a feature request to move this function to the public API (see Microsoft/TypeScript/issues/25225). After approving the request, I will be able to change the code to use this function and the issue will be fixed.

AFAIK this function was added in `2.6.x` TypeScript's version, so we need to release the version with dropping support `< 2.6.x` versions.
